### PR TITLE
New tool: case-study extender

### PR DESCRIPTION
**Pull request by vulder**
Implemented new case-study extender tool that allows users to adapt and extend existing case-studies.
Case studies now have different stages, i.e., partitions of revisions that logically group revisions. 

resolves se-passau/VaRA#401

### CODE REVIEW

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 5
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  5
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/tools/commit_map.py  1
- varats/driver.py  8
         

Complexity decreasing per file
==============================
+ varats/paper/paper_config.py  -1
+ varats/paper/case_study.py  -6
         

Clones added
============
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

File: varats/paper/case_study.py [hunk 1]
```diff
@@ -89,34 +135,77 @@ def revisions(self):
...
+                          revisions: [(str, int)],
+                          stage_num=0,
+                          sort_revs=True):
         """
         Add multiple revisions to this case study.
 
         Args:
-            revisions: List of tuples with commit_hash and id
+            stage: The stage to insert the revisions
+            revisions: List of triples with (commit_hash, id) to be inserted
```
Comment (julianbreiteneicher) [inline varats/paper/case_study.py:201]: triple or tuple?

File: varats/driver.py [hunk 2]
```diff
@@ -374,25 +429,58 @@ def main_casestudy():
         if 'paper_config' in args:
```
Comment (julianbreiteneicher) [inline varats/driver.py:434]: `args['subcommand']` crashes if I start `vara-cs` without any params:

```
(vara-tool-suite) [jb@scotty ~/work/hiwi:master*] vara-cs
Traceback (most recent call last):
  File "/home/jb/opt/virtualenvs/vara-tool-suite/bin/vara-cs", line 11, in <module>
    load_entry_point('VaRA-Tool-Suite', 'console_scripts', 'vara-cs')()
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/driver.py", line 428, in main_casestudy
    if args['subcommand'] == 'status':
KeyError: 'subcommand'
```

I suggest to print the output of '-h' in this case instead.

File: varats/driver.py [hunk 3]
```diff
@@ -340,30 +350,75 @@ def main_casestudy():
...
+    ext_parser.add_argument(
+        "--plot-type",
+        action=enum_action(PlotTypes),
+        help="Plot to calculate new revisions from.")
+    ext_parser.add_argument(
+        "--result-folder",
+        help="Maximal expected gradient in percent between two revisions")
+    add_common_args(ext_parser)
 
     args = {
```
Comment (julianbreiteneicher) [inline varats/driver.py:424]: The command `vara-cs ext -h` crashes:

```
Traceback (most recent call last):
  File "/home/jb/opt/virtualenvs/vara-tool-suite/bin/vara-cs", line 11, in <module>
    load_entry_point('VaRA-Tool-Suite', 'console_scripts', 'vara-cs')()
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/driver.py", line 425, in main_casestudy
    for k, v in vars(parser.parse_args()).items() if v is not None
  File "/usr/lib/python3.7/argparse.py", line 1749, in parse_args
    args, argv = self.parse_known_args(args, namespace)
  File "/usr/lib/python3.7/argparse.py", line 1781, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.7/argparse.py", line 1969, in _parse_known_args
    positionals_end_index = consume_positionals(start_index)
  File "/usr/lib/python3.7/argparse.py", line 1946, in consume_positionals
    take_action(action, args)
  File "/usr/lib/python3.7/argparse.py", line 1855, in take_action
    action(self, namespace, argument_values, option_string)
  File "/usr/lib/python3.7/argparse.py", line 1152, in __call__
    subnamespace, arg_strings = parser.parse_known_args(arg_strings, None)
  File "/usr/lib/python3.7/argparse.py", line 1781, in parse_known_args
    namespace, args = self._parse_known_args(args, namespace)
  File "/usr/lib/python3.7/argparse.py", line 1987, in _parse_known_args
    start_index = consume_optional(start_index)
  File "/usr/lib/python3.7/argparse.py", line 1927, in consume_optional
    take_action(action, args, option_string)
  File "/usr/lib/python3.7/argparse.py", line 1855, in take_action
    action(self, namespace, argument_values, option_string)
  File "/usr/lib/python3.7/argparse.py", line 1037, in __call__
    parser.print_help()
  File "/usr/lib/python3.7/argparse.py", line 2474, in print_help
    self._print_message(self.format_help(), file)
  File "/usr/lib/python3.7/argparse.py", line 2458, in format_help
    return formatter.format_help()
  File "/usr/lib/python3.7/argparse.py", line 284, in format_help
    help = self._root_section.format_help()
  File "/usr/lib/python3.7/argparse.py", line 215, in format_help
    item_help = join([func(*args) for func, args in self.items])
  File "/usr/lib/python3.7/argparse.py", line 215, in <listcomp>
    item_help = join([func(*args) for func, args in self.items])
  File "/usr/lib/python3.7/argparse.py", line 215, in format_help
    item_help = join([func(*args) for func, args in self.items])
  File "/usr/lib/python3.7/argparse.py", line 215, in <listcomp>
    item_help = join([func(*args) for func, args in self.items])
  File "/usr/lib/python3.7/argparse.py", line 525, in _format_action
    help_text = self._expand_help(action)
  File "/usr/lib/python3.7/argparse.py", line 614, in _expand_help
    return self._get_help_string(action) % params
```

File: varats/paper/case_study.py [hunk 4]
```diff
@@ -89,34 +135,77 @@ def revisions(self):
...
                 return True
 
         return False
 
-    def include_revision(self, revision, commit_id):
+    def has_revision_in_stage(self, revision: str, num_stage) -> bool:
+        """
+        Check if a revision of a specific stage.
+        """
+        if self.num_stages <= num_stage:
```
Comment (julianbreiteneicher) [inline varats/paper/case_study.py:169]: Crash for `vara-cs ext paper_configs/test/gzip_0.case_study distrib_add benchbuild/tmp/gzip-HEAD --distribution uniform --num-rev 5 --merge-stage 1`:

```
Traceback (most recent call last):
  File "/home/jb/opt/virtualenvs/vara-tool-suite/bin/vara-cs", line 11, in <module>
    load_entry_point('VaRA-Tool-Suite', 'console_scripts', 'vara-cs')()
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/driver.py", line 468, in main_casestudy
    extend_case_study(case_study, cmap, args['strategy'], **args)
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/paper/case_study.py", line 364, in extend_case_study
    extend_with_distrib_sampling(case_study, cmap, **kwargs)
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/plots/plot_utils.py", line 28, in wrapper_func
    return func(*args, **kwargs)
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/paper/case_study.py", line 394, in extend_with_distrib_sampling
    key=lambda x: x[1]) if not case_study.
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/paper/case_study.py", line 395, in <listcomp>
    has_revision_in_stage(rev_item[0], kwargs['merge_stage'])
  File "/home/jb/work/hiwi/repos/VaRA-Tool-Suite/varats/paper/case_study.py", line 169, in has_revision_in_stage
    if self.num_stages <= num_stage:
TypeError: '<=' not supported between instances of 'int' and 'str'
```

**Review by julianbreiteneicher — CHANGES_REQUESTED** [review]
Looks good so far, but there are still a few issues.

One additional thing that I just noticed:
We recently talked about the consistency of the command line parameters.
More specifically, you said that a parameter with a specific name should always mean the same thing.
But I think with `num_rev` this is not completely the case:
When I create a case-study then it says how many revisions should be in the resulting study in total.
However, when extending a case-study it says how many revisions should be added (and not how many will exist after the extension).
Is my thinking correct or am I misunderstanding something here?

File: varats/driver.py (hunk 3 above)
Comment (vulder) [reply varats/driver.py:424]: python what are you doing:
```
        help="Maximal expected gradient in percent between " +
        "two revisions, e.g., 5 for 5%")
```
help is know a format string... guess what the correct escape is...
```
        help="Maximal expected gradient in percent between " +
        "two revisions, e.g., 5 for 5%%")
```
double %
*face -> wall*

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
           

Complexity increasing per file
==============================
- varats/driver.py  9
         

Complexity decreasing per file
==============================
+ varats/paper/case_study.py  -6
+ varats/paper/paper_config.py  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3636280&bid=12867868)

**Review by julianbreiteneicher — APPROVED** [review]
Looks good now! :+1:

**Pull request merged**